[TextureMapper] Preserve 3d may split layers unnecessarily when they don't intersect

zhani opened the pull request:
#### ca6c3486ddd32d840ebdaee645aa025d93090955
<pre>
[TextureMapper] Preserve 3d may split layers unnecessarily when they don&apos;t intersect
<a href="https://bugs.webkit.org/show_bug.cgi?id=284250">https://bugs.webkit.org/show_bug.cgi?id=284250</a>

Reviewed by Fujii Hironori.

Add preprocessing to conditionally skip BSP tree building for 3D rendering

Introduce a preprocessing step to determine if BSP tree construction can be
bypassed. If the 3D scene lacks intersecting layers or has an unambiguous
rendering order, building the BSP tree is unnecessary. Skipping BSP tree
layer splits enhances performance, especially on devices with slow stencil
performance like RPi3.

This patch implements:
- **Broad Phase:** Sweep-and-prune algorithm to identify potential intersecting layers.
- **Narrow Phase:** Projects potential collision pairs onto the ZY plane for intersection checks.

The ZY projection effectively detects both intersections and ambiguous rendering
orders, ensuring that BSP tree building is only performed when necessary.

* Source/WebCore/platform/graphics/texmap/ClipPath.h:
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeFlattenedRegion const):
(WebCore::TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective const):
(WebCore::TextureMapperLayer::collect3DRenderingContextLayers):
(WebCore::TextureMapperLayer::effectiveLayerRect const):
(WebCore::TextureMapperLayer::computeFlattenedRegion): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::isLeafOf3DRenderingContext const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
(WebCore::projectPolygonToZYPlane):
(WebCore::edgeNormal):
(WebCore::projectQuadOnAxis):
(WebCore::quadsIntersect):
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::computeBoundingBox):
(WebCore::TextureMapperLayer3DRenderingContext::sweepAndPrune):
(WebCore::TextureMapperLayer3DRenderingContext::buildTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTree):
(WebCore::TextureMapperLayer3DRenderingContext::classifyLayer):
(WebCore::TextureMapperLayer3DRenderingContext::classifyPolygon): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:

Canonical link: <a href="https://commits.webkit.org/288080@main">https://commits.webkit.org/288080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d354a60932c1e8fe1357e366c3bf6c772e822733

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86374 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32822 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9173 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63811 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21543 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84896 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1014 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74461 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44097 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/913 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28636 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31275 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29243 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87809 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9066 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6439 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72162 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9251 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70282 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71394 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15492 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14413 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12674 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9017 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14549 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8858 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12381 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/texmap/FloatBox3D.h:
```diff
@@ -0,0 +1,61 @@
+            (m_min.x() + m_max.x()) / 2.0f,
+            (m_min.y() + m_max.y()) / 2.0f,
+            (m_min.z() + m_max.z()) / 2.0f
+        );
```
~~~
return (m_min + m_max) * 0.5f;
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/FloatBox3D.h:
```diff
@@ -0,0 +1,61 @@
+    FloatBox3D(const FloatPoint3D& min, const FloatPoint3D& max);
+
+    const FloatPoint3D&  min() const { return m_min; }
+    const FloatPoint3D&  max() const { return m_max; }
```
single space after `&`.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
 
+void FloatPolygon3D::computeBounds()
+{
+    FloatPoint3D minCorner =  vertexAt(0);
```
single space after `=`.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -44,38 +46,51 @@ class TextureMapperLayer3DRenderingContext final {
+    struct Layer final {
         FloatPolygon3D geometry;
-        TextureMapperLayer* layer = { nullptr };
+        TextureMapperLayer* textureMapperLayer = { nullptr };
```
WebKit doesn't put `=` here. `textureMapperLayer { nullptr };`

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
-                    clipVertices.append(v.y());
-                }
+    Vector<Layer> layers;
+    for (size_t i = 0; i < textureMapperLayers.size(); ++i) {
```
the original code `for (auto* layer : textureMapperLayers)` looks nice. Why do you want to use `i` here?

zhani replied:
No reason, thanks. Leftover from some trials I had for debugging. Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
+
+            auto [iMin, iMax] = layerBounds(layers[firstIndex], axis);
+            auto [jMin, jMax] = layerBounds(layers[secondIndex], axis);
+            if (jMin > iMax)
```
I think `if (jMin >= iMax)` is better.

zhani replied:
Good point. Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
+    bool hasIntersections = false;
+    if (potentialIntersections.size() > 0) {
```
if (!potentialIntersections.isEmpty())

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
-                for (size_t i = 0; i < xy.size(); i += 2)
-                    points.append(FloatPoint(xy.at(i), xy.at(i + 1)));
+            ClipPath clipPath(WTFMove(points), 0, 0);
+            paintLayerFunction(layer.textureMapperLayer, clipPath);
```
ClipPath has the default ctor. This code can be
~~~
paintLayerFunction(layer.textureMapperLayer, { });
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
+        // Sort back to front
+        std::sort(layers.begin(), layers.end(), [](const Layer& layerA, const Layer& layerB) {
+            auto centerA = layerA.geometry.bounds().center();
+            auto centerB = layerB.geometry.bounds().center();
```
I think center is not the correct way to order layers.
For example, in the diagram, layer A, B, C are drawn by sorted z order.
But, they should be drawn in the oder layer B, A, C.

<img width="702" alt="{BE5E2E4D-82D8-4FB3-BE42-FEA6E906E0EF}" src="https://github.com/user-attachments/assets/c61fef84-ab01-4d8c-a120-31913ccc0f7b" />

zhani replied:
So Y axis is pointing towards viewer here so projections on zy will intersect, right? In that case, this goes to BSP.

fujii replied:
You are right. Why do you compare centers here? I think comparing min Z is enough. Right?

zhani replied:
you are right, comparing to min Z should be fine

fujii replied:
Then, `layerA.geometry.bounds().center()` → `layerA.geometry.bounds().min().z()`

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
+        potentialIntersectionsXY.begin(), potentialIntersectionsXY.end(),
+        potentialIntersectionsZAxis.begin(), potentialIntersectionsZAxis.end(),
+        std::inserter(potentialIntersections, potentialIntersections.begin())
+    );
```
I think this is inefficient.
How about this?
* Adding `FloatBox3D::intersects(const FloatBox3D&)` method like `FloatRect::intersects`.
* Checking if each pair of layers intersect

zhani replied:
I was a bit too hasty in pushing my code. I accidentally pushed debug branch that had some extra. Code is now updated. Doing sweep and prune on all axis is effectively same as doing bounding box intersection test. However, my idea is to do sweep and prune only on x-axis. This reduces the set needed for zy intersection tests. Updated the code to do only x axis sweep and prune

fujii replied:
Why only on x-axis? bounding box intersection test is pretty cheep.

zhani replied:
Yes, bounding box intersection tests are inexpensive. However, using them here would result in duplicate tests since the zy intersection already accounts for bounding box scenarios along the z and y dimensions. Additionally, full bounding box tests would involve testing each layer against every other layer, leading to redundant computations. An x-axis sweep-and-prune efficiently reduces the number of pairwise tests. That said, for typical scene sizes, the performance gain over full bounding box tests might be negligible.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,300 @@
+    );
+
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
```
I think this comment is not correct. This is approximation.
For example, in the diagram below, layer A, B, C don't intersect.
However, in YZ plane, they intersect.

zhani replied:
Yes, they intersect in ZY plane and in that case BSP will handle it. This is on purpose, because it's very difficult to cover all non-intersecting scenarios without BSP. Some of non-intersecting cases need BSP because of ambiguous rendering order like in the test case attached to bug. If ZY plane intersection doesn't happen then we can be sure that  layers can be rendered without clip in z-order, otherwise BSP will handle it.
 In this scenario BSP doesn't do splitting either because split planes are coplanar to layers. BSP detects that layers are fully behind each other and renders them without clip.

fujii commented on Source/WebCore/platform/graphics/texmap/ClipPath.h:
```diff
@@ -37,20 +37,21 @@ namespace WebCore {
 public:
-    ClipPath(Vector<FloatPoint>&& vertices, unsigned bufferID, unsigned bufferOffsetInBytes);
+    ClipPath() = default;
+    ClipPath(Vector<FloatPoint>&& vertices, unsigned bufferID , unsigned bufferOffsetInBytes);
```
You added a space after `bufferID`.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -44,38 +46,51 @@ class TextureMapperLayer3DRenderingContext final {
-    void buildTree(TextureMapperLayerNode&, Deque<TextureMapperLayerPolygon>&);
-    void traverseTree(TextureMapperLayerNode&, const std::function<void(TextureMapperLayerNode&)>&);
-    static PolygonPosition classifyPolygon(const TextureMapperLayerPolygon&, const FloatPlane3D&);
+    using SweepAndPrunePairs = StdSet<std::pair<size_t, size_t>>;
```
Why not HashSet? I think we should use HashSet unless we have a special reason.

zhani replied:
Only purpose is to ensure we don't have duplicate entries. Changed to HashSet

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,277 @@
+    axes[6] = edgeNormal(quadB.p3(), quadB.p4());
+    axes[7] = edgeNormal(quadB.p4(), quadB.p1());
+
+    for (int i = 0; i < 8; ++i) {
```
How about `for (auto& axis : axes) {` or `std::all_of`?

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -44,38 +46,51 @@ class TextureMapperLayer3DRenderingContext final {
+    bool quadsIntersect(const FloatQuad&, const FloatQuad&) const;
+    void buildTree(LayerNode&, Deque<Layer>&);
+    void traverseTree(LayerNode&, const std::function<void(LayerNode&)>&);
+    static LayerPosition classifyLayer(const Layer&, const FloatPlane3D&);
```
I think we should add `static` to member functions which don't need `this`.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,277 @@
+    auto p2 = layer.geometry.vertexAt(1);
+    auto p3 = layer.geometry.vertexAt(2);
+    auto p4 = layer.geometry.vertexAt(3);
+    return FloatQuad(
```
WebKit doesn't write `FloatQuad(` here.
~~~
  return {
     ... 
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,277 @@
+        case SweepAndPruneAxis::YAxis:
+            return { layer.geometry.bounds().min().y(), layer.geometry.bounds().max().y() };
+        default:
+            return { layer.geometry.bounds().min().z(), layer.geometry.bounds().max().z() };
```
You can remove code other than XAxis.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/FloatBox3D.h:
```diff
@@ -0,0 +1,57 @@
+    const FloatPoint3D& min() const { return m_min; }
+    const FloatPoint3D& max() const { return m_max; }
+
+    FloatPoint3D center() const
```
You can remove.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -44,38 +46,51 @@ class TextureMapperLayer3DRenderingContext final {
     };
 
-    struct TextureMapperLayerPolygon final {
+    enum class SweepAndPruneAxis {
```
You can remove SweepAndPruneAxis.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
+
+        maxCorner.setX(std::max(maxCorner.x(), point.x()));
+        maxCorner.setY(std::max(maxCorner.y(), point.y()));
+        maxCorner.setZ(std::max(maxCorner.z(), point.z()));
```
IIUC, you don't use these min of Y and Z. Can you remove them?

zhani replied:
Then bounds wouldn't be bounding box. I think it's better to leave them so that bounds is actually the bounding box of 3d polygon. It's easier to understand. Otherwise function name should be changed to xRange or something like that

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,266 @@
+            auto [iMin, iMax] = layerBounds(layers[firstIndex]);
+            auto [jMin, jMax] = layerBounds(layers[secondIndex]);
+            if (jMin >= iMax)
+                break; // No overlap possible
```
How about adding bounding box intersection check here?

~~~c++
iMin = layers[firstIndex].geometry.bounds().min();
iMax = layers[firstIndex].geometry.bounds().max();
jMin = layers[secondIndex].geometry.bounds().min();
jMax = layers[secondIndex].geometry.bounds().max();
if (iMin.x() <= jMax.x())
    break;
if (iMin.y() >= jMax.y() || iMax.y() <= jMin.y())
    continue;
if (iMin.z() >= jMax.z() || iMax.z() <= jMin.z())
    continue;
~~~

zhani replied:
`if (iMin.x() <= jMax.x())` is wrong, but I think it's a typo and you probably meant `if (iMin.x() >= jMax.x())`.
But otherwise I think this should work. I cannot come up with any scenario why it wouldn't work.
I will run some more tests a bit later. But in any case zy -projection tests are still needed after sweep even with this code applied

zhani replied:
This is now applied

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,266 @@
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
+    bool hasIntersections = false;
+    if (!potentialIntersections.isEmpty()) {
+        for (const auto& p : potentialIntersections) {
```
I don't think we need to check if it's empty, the loop won't iterate anyway.

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,266 @@
-                auto xy = clipVertices.subvector(polygon.clipVertexBufferOffset, numberOfClipVertices * 2);
-                for (size_t i = 0; i < xy.size(); i += 2)
-                    points.append(FloatPoint(xy.at(i), xy.at(i + 1)));
+        for (auto& layer : layers)
+            paintLayerFunction(layer.textureMapperLayer, { });
```
We could return after this so that we don't need the else and we save an indentation for the code below.

zhani replied:
Fixed

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,266 @@
+        Deque<Layer> layerDeque;
+        for (auto& layer : layers)
+            layerDeque.append(WTFMove(layer));
+        layers.clear();
```
Can we use a Deque for the layers and then we don't need to convert the Vector to Deque here?

zhani replied:
sweepAndPruneXAxis operates with indices and returns index pairs. With Vector it's fast to access index directly. However, code has already change quite a bit from original approach and sweepAndPruneXAxis can be probably implemented a bit differently so that Deque can be used for the layers

zhani replied:
Seem the first part of the algorithm cannot be done with Deque. SweepAndPruneXAxis needs sort layers and thus random access iterator. Deque doesn't provide such.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,118 +29,266 @@
+}
+
+TextureMapperLayer3DRenderingContext::SweepAndPrunePairs TextureMapperLayer3DRenderingContext::sweepAndPruneXAxis(
+    const Vector<Layer>& layers)
```
```suggestion
TextureMapperLayer3DRenderingContext::SweepAndPrunePairs TextureMapperLayer3DRenderingContext::sweepAndPruneXAxis(const Vector<Layer>& layers)
```

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatBox3D.h:
```diff
@@ -0,0 +1,52 @@
+    WTF_MAKE_TZONE_ALLOCATED(FloatBox3D);
+public:
+    FloatBox3D() = default;
+    FloatBox3D(const FloatPoint3D& min, const FloatPoint3D& max);
```
I would implement all methods inline and avoid the .cpp file.

zhani replied:
FloatBox3D removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatBox3D.h:
```diff
@@ -0,0 +1,52 @@
+    void set(const FloatPoint3D& min, const FloatPoint3D& max);
+
+private:
+
```
Superfluous newline.

zhani replied:
FloatBox3D removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
 
+void FloatPolygon3D::computeBounds()
+{
+    FloatPoint3D minCorner = vertexAt(0);
```
auto?

zhani replied:
FloatPolygon3D::computeBounds() removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
+    computeBounds();
 }
 
+void FloatPolygon3D::computeBounds()
```
Why don't we compute bounds on-demand? Any reason to store it?

nikolaszimmermann replied:
I saw that it's probably used more than once per FloatPolygon3D -- we could compute on-demand upon `bounds()` access and store it in a `mutable FloatRect m_bounds`.

zhani replied:
FloatPolygon3D::computeBounds() removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -29,40 +29,83 @@
+
+    Vector<Layer> layers;
+    for (auto* layer : textureMapperLayers)
+        layers.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
```
That is hard to read -- one can only guess what objects you are constructing here. The spurious boolean in the end raises eyebrows. `layers.append(Layer::createFoo(layer))` , and have  static createFoo method encapsulating the logic in `Layer`?

zhani replied:
Didn't add create function but changed it to be more readable.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -75,72 +118,175 @@ void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, c
+// https://en.wikipedia.org/wiki/Hyperplane_separation_theorem
+bool TextureMapperLayer3DRenderingContext::quadsIntersect(const FloatQuad& quadA, const FloatQuad& quadB)
+{
+    FloatPoint axes[8];
```
std::array<FloatPoint, 8> axes?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
```diff
@@ -44,38 +46,45 @@ class TextureMapperLayer3DRenderingContext final {
+    using SweepAndPrunePairs = HashSet<std::pair<size_t, size_t>>;
+
+    static SweepAndPrunePairs sweepAndPruneXAxis(const Vector<Layer>&);
+    static FloatQuad projectLayerToZYPlane(const Layer&);
```
I'd keep all those helpers inlined in the .cpp file, instead of declaring them here.

zhani replied:
Moved all functions that don't use private structs to .cpp static inlines

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/ClipPath.h:
```diff
@@ -37,20 +37,21 @@ namespace WebCore {
     const Vector<FloatPoint> m_vertices;
-    unsigned m_bufferID;
-    unsigned m_bufferOffsetInBytes;
+    unsigned m_bufferID { 0 };
```
In general it's good practice to separate the trivial style fixes in a PR that preceeds this one -- you can keep them here, just wanted to mention it.

zhani replied:
Ok, thanks for the hint

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
+    FloatPoint3D maxCorner = vertexAt(0);
+
+    unsigned numVertices = numberOfVertices();
+    for (unsigned i = 1; i < numVertices; i++) {
```
s/i++/++i/

zhani replied:
FloatPolygon3D::computeBounds() removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
```diff
@@ -49,14 +51,38 @@ FloatPolygon3D::FloatPolygon3D(const FloatRect& rect, const TransformationMatrix
+    FloatPoint3D minCorner = vertexAt(0);
+    FloatPoint3D maxCorner = vertexAt(0);
+
+    unsigned numVertices = numberOfVertices();
```
No need for the local variable, all modern compilers optimize  away the `numberOfVertices()` call in the loop.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h:
```diff
@@ -49,13 +51,18 @@ class FloatPolygon3D {
+    void computeBounds();
+
     Vector<FloatPoint3D> m_vertices;
     FloatPoint3D m_normal = { 0, 0, 1 };
```
s/= {/ {/

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -75,72 +118,175 @@ void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, c
-// Build BSP tree for rendering polygons with painter's algorithm.
+TextureMapperLayer3DRenderingContext::SweepAndPrunePairs TextureMapperLayer3DRenderingContext::sweepAndPruneXAxis(const Vector<Layer>& layers)
+{
+    auto layerBoundsXRange = [](const Layer& layer) -> std::pair<float, float> {
```
This could be an helpful accessor for Layer directly?

zhani replied:
not needed anymore as such, Created separate private BoundingBox object

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
```diff
@@ -30,39 +30,150 @@
 #include "TextureMapperLayer.h"
-#include <wtf/Deque.h>
+#include <numeric>
+#include <wtf/StdSet.h>
```
Remove this line. StdSet is no longer used.

zhani replied:
Fixed

fujii reviewed (approved):
LGTM. Great improvement!

zhani committed:
[TextureMapper] Preserve 3d may split layers unnecessarily when they don't intersect
https://bugs.webkit.org/show_bug.cgi?id=284250

Reviewed by Fujii Hironori.

Add preprocessing to conditionally skip BSP tree building for 3D rendering

Introduce a preprocessing step to determine if BSP tree construction can be
bypassed. If the 3D scene lacks intersecting layers or has an unambiguous
rendering order, building the BSP tree is unnecessary. Skipping BSP tree
layer splits enhances performance, especially on devices with slow stencil
performance like RPi3.

This patch implements:
- **Broad Phase:** Sweep-and-prune algorithm to identify potential intersecting layers.
- **Narrow Phase:** Projects potential collision pairs onto the ZY plane for intersection checks.

The ZY projection effectively detects both intersections and ambiguous rendering
orders, ensuring that BSP tree building is only performed when necessary.

* Source/WebCore/platform/graphics/texmap/ClipPath.h:
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.cpp:
* Source/WebCore/platform/graphics/texmap/FloatPolygon3D.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeFlattenedRegion const):
(WebCore::TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective const):
(WebCore::TextureMapperLayer::collect3DRenderingContextLayers):
(WebCore::TextureMapperLayer::effectiveLayerRect const):
(WebCore::TextureMapperLayer::computeFlattenedRegion): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::isLeafOf3DRenderingContext const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
(WebCore::projectPolygonToZYPlane):
(WebCore::edgeNormal):
(WebCore::projectQuadOnAxis):
(WebCore::quadsIntersect):
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::computeBoundingBox):
(WebCore::TextureMapperLayer3DRenderingContext::sweepAndPrune):
(WebCore::TextureMapperLayer3DRenderingContext::buildTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTree):
(WebCore::TextureMapperLayer3DRenderingContext::classifyLayer):
(WebCore::TextureMapperLayer3DRenderingContext::classifyPolygon): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:

Canonical link: https://commits.webkit.org/288080@main
